### PR TITLE
Parameterize more of raftstore over engines

### PR DESCRIPTION
**Pull request by brson**
### What problem does this PR solve?

Part of https://github.com/tikv/tikv/issues/6402

Progress toward making raftstore work with alternate engines.

### What is changed and how it works?

This PR is focused on turning RocksEngine types in the raftstore "peer" and "msg" modules into type parameters. Many types are now parameterized over two typarams, EK (the "kv" engine), and ER (the "raft" engine).

This work is clearly incomplete and there are many obvious next-steps, but there is so much churn here that I want to get it landed before continuing.

Type-parameterizion is falling into place quickly now, and the path is pretty clear to getting all of raftstore abstracted over the two engines it needs.

### Check List 

Type checking and unit testing.

### Release note 

- No release note.

### CODE REVIEW

File: components/raftstore/src/store/fsm/store.rs [hunk 1]
```diff
@@ -150,12 +150,19 @@ impl StoreMeta {
     }
 }
 
-pub struct RaftRouter<S: Snapshot> {
-    pub router: BatchRouter<PeerFsm<S>, StoreFsm>,
+pub struct RaftRouter<EK, ER, S>
```
Comment (5kbpers) [inline components/raftstore/src/store/fsm/store.rs:153]: Could we use `Ek::Snapshot` instead of `S`?

Comment (Connor1996) [reply components/raftstore/src/store/fsm/store.rs:153]: +1

Comment (brson) [reply components/raftstore/src/store/fsm/store.rs:153]: Yes. That is part of the intended future work. I will start on that patch now, but ask that you consider landing this first.

**Comment by brson** [comment]
Updated to master.

**Comment by brson** [comment]
@5kbpers @Connor1996 the extra `S` parameter is removed now.

**Pull request merged**